README.pod
=head1 NAME

Games::FrogJump - An ASCII game for fun

=head1 SYNOPIS

 use Games::FrogJump;
 Games::FrogJump->new->run;

=head1 DESCRIPTION

This module is an ASCII game , It runs at command-line. Control the frogs jump to each side.

Play the game with command:

  frogjump

=begin HTML

<p><img src="http://ww3.sinaimg.cn/large/a8e976cegw1ej24klqiqkj20gf07gt8t.jpg" /></p>

=end HTML

=head1 TODO

=over

=item * Add some tip when can not jump

=item * Add an animation

=back

=head1 ACKNOWLEDGEMENTS

BLAIZER, Author of L<Games::2048>

=head1 AUTHOR

tadegenban <tadegenban@gmail.com>

=cut


Changes
{{$NEXT}}

blablabla sdsd s v0.02

{{$NEXT}}

sdsdsd v0.01


lib/Games/FrogJump.pm
=head1 NAME

Games::FrogJump - An ASCII game for fun

=head1 SYNOPIS

 use Games::FrogJump;
 Games::FrogJump->new->run;

=head1 DESCRIPTION

This module is an ASCII game , It runs at command-line. Control the frogs jump to each side.

Play the game with command:

  frogjump

=begin HTML

<p><img src="http://ww3.sinaimg.cn/large/a8e976cegw1ej24klqiqkj20gf07gt8t.jpg" /></p>

=end HTML

=head1 AUTHOR

tadegenban <tadegenban@gmail.com>

=cut

package Games::FrogJump;
use 5.012;
use Moo;

our $VERSION = '0.04';

use Time::HiRes;
our $FRAME_TIME = 1/20;

use Games::FrogJump::Game;
use Games::FrogJump::Frog;
use Games::FrogJump::Input;
use Games::FrogJump::Animation;

sub run {
    my $self = shift;
    my $game;
    my $restart;
    my $quit;

    while (!$quit) {

        if ( !$game ) {
            $game = Games::FrogJump::Game->new();
        }
        if ( $restart ){
            $game->restart;
        }
        $game->init;
      RUN: $game->draw;
        my $time = Time::HiRes::time;
      PLAY: while ( 1 ) {
          while ( defined(my $key = Games::FrogJump::Input::read_key) ) {
              my $cmd = Games::FrogJump::Input::key_to_cmd($key);
              if ( $cmd eq 'quit' ){
                  $quit = 1;
                  last PLAY;
              }
              if ( $cmd eq 'restart' ){
                  $restart = 1;
                  last PLAY;
              }
              if ( $cmd ) {
                  $game->act($cmd);
              }
          }
          if ( @{$game->animations} ){
              foreach my $animation ( @{$game->animations} ){
                  $game->remove_animation($animation) if $animation->end;
                  $animation->update;
              }
          }
          $game->draw;
          my $new_time = Time::HiRes::time;
          my $delta_time = $new_time - $time;
          my $delay = $FRAME_TIME - $delta_time;
          $time = $new_time;
          if ($delay > 0) {
              Time::HiRes::sleep($delay);
              $time += $delay;
          }
          if ( $game->win || $game->lose ){
              last PLAY;
          }
      }
        if ( $game->win ){
            $game->draw_win;
            $quit = 1;
        }
    }
    $game->draw_quit;
}
1;


lib/Games/FrogJump/Frog.pm
package Games::FrogJump::Frog;
use 5.012;
use Moo;
use if $^O eq "MSWin32", "Win32::Console::ANSI";
use Term::ANSIColor;

has direction    => is => 'ro', default => 'null';
has stone_index  => is => 'rw', default => -1;
has width        => is => 'rw', default => 3;
has bg_color     => is => 'rw', default => 'reset';
has fr_color     => is => 'rw', default => 'blue';
has stored_ansi  => is => 'rw';
has ansi         => is => 'rw', default => '   ';
has x            => is => 'rw', default => 0;
has y            => is => 'rw', default => 0;

sub graph {
    my $self = shift;
    return color( $self->fr_color . ' ' . $self->bg_color ) . $self->ansi . color('reset');
}

sub active {
    my $self = shift;
    $self->fr_color('red');
    $self->stored_ansi($self->ansi);
    $self->ansi('@@_') if $self->direction eq 'left';
    $self->ansi('_@@') if $self->direction eq 'right';
}

sub unactive {
    my $self = shift;
    $self->fr_color('blue');
    $self->ansi($self->stored_ansi);
}

1;


lib/Games/FrogJump/Board.pm
package Games::FrogJump::Board;
use 5.012;
use strict;
use warnings;
use Moo;

use Text::Wrap;
use if $^O eq "MSWin32", "Win32::Console::ANSI";
use Term::ANSIColor;
use List::Util qw/max min/;
use List::MoreUtils qw/none/;
use Color::ANSI::Util qw/ansifg ansibg/;

has frog_number   => is => 'rw', default => 6;
has frog_width    => is => 'rw', default => 3;
has stone_number  => is => 'rw', default => sub { my $self = shift; $self->frog_number + 1 };
has stone_width   => is => 'rw', default => 7;
has stone_gap     => is => 'rw', default => 2;

has border_width  => is => 'rw', default => 2;
has border_height => is => 'rw', default => 1;
has border_color  => is => 'rw', default => 'on_cyan';
has content_width => is => 'rw', default => 70;
has content_height => is => 'rw', default => 13;
has padding       => is => 'rw', default => 4;

has _frogs       => is => 'lazy';
has current_frog => is => 'rw', default => '';
has animations   => is => 'rw', default => sub { [] };

sub draw {
    my ( $self ) = @_;
    local $| = 1;
    $self->hide_cursor;
    $self->draw_frog;
    $self->draw_border;
    $self->draw_title;
    $self->draw_guide;
    $self->draw_stone;

}

sub draw_frog {
    my $self = shift;
    foreach my $frog ( @{$self->_frogs} ){
        if ( $frog->x ne $frog->oldx or $frog->y ne $frog->oldy ){
            $self->save_cursor;
            $self->move_cursor($frog->oldx, $frog->oldy);
            print ' ' x 3;
            $self->restore_cursor;

            $self->save_cursor;
            $self->move_cursor($frog->x, $frog->y);
            print $frog->graph;
            $self->restore_cursor;
        }
        else{
            $self->save_cursor;
            $self->move_cursor($frog->x, $frog->y);
            print $frog->graph;
            $self->restore_cursor;
        }
    }
}

sub draw_stone {
    my ( $self )  = @_;
    $self->save_cursor;
    $self->move_cursor(6, $self->content_height);
    foreach my $index ( 0..$self->stone_number-1 ){
        my $stone;
        my $gap;
        $stone = ' ' x $self->stone_width;
        $stone = color('reverse cyan'). $stone . color('reset');
        $gap   = ' ' x $self->stone_gap;
        print $stone;
        print $gap;
    }
    $self->restore_cursor;
}

sub draw_guide {
    my $self = shift;
    $self->save_cursor;
    $self->move_cursor($self->content_width-15, 2);
    print color('green') . '<-' . color('reset') . ' : select frog';
    $self->move_cursor(-16, 1);
    print color('green') . '->' . color('reset') . ' : select frog';
    $self->move_cursor(-16, 1);
    print color('green') . 'sp' . color('reset') . ' : jump       ';
    $self->move_cursor(-16, 1);
    print color('green') . 'r ' . color('reset') . ' : restart    ';
    $self->move_cursor(-16, 1);
    print color('green') . 'q ' . color('reset') . ' : quit       ';
    $self->restore_cursor;
}

sub draw_title {
    my $self = shift;
    $self->save_cursor;
    $self->move_cursor(30, 2);
    print color('green') . 'Frog Jump --_' . color('reset');
    $self->restore_cursor;
}

sub draw_win {
    my $self = shift;
    $self->save_cursor;
    $self->move_cursor(30, 2);
    print color('green') . 'You  Win  @@_' . color('reset');
    $self->restore_cursor;
}

sub draw_quit {
    my $self = shift;
    $self->move_cursor(0, $self->content_height + 5);
    say color('reset') . '';
    $self->show_cursor;
}
sub draw_border {
    my $self = shift;
    $self->save_cursor;
    say color($self->border_color), " " x $self->board_width, color("reset") for 1..$self->border_height;
    foreach my $col ( 0..$self->content_height-1 ){
        print color($self->border_color), " " x $self->border_width, color("reset");
        $self->move_cursor(70, 0);
        print color($self->border_color), " " x $self->border_width, color("reset");
        print "\n";
    }

    say color($self->border_color), " " x $self->board_width, color("reset") for 1..$self->border_height;
    $self->restore_cursor;
}

sub add_animation {
    my ( $self, $animation ) = @_;
    if ( none { $_->name eq $animation->name } @{$self->animations} ){
        push $self->animations, $animation;
    }
}

sub remove_animation {
    my ( $self, $animation ) = @_;
    my $new_animations = [ grep { $_->name ne $animation->name } @{$self->animations} ];
    $self->animations($new_animations);
}

sub board_width {
    my $self = shift;
    return $self->content_width + $self->border_width * 2;
}

sub board_height {
    my $self = shift;
    return $self->border_height * 2 + $self->content_height;
}

sub move_cursor {
    my ( $self, $dx, $dy ) = @_;
    $dx > 0 ? do { printf "\e[%dC", $dx } : $dx < 0 ? do { printf "\e[%dD", -$dx } : do {};
    $dy > 0 ? do { printf "\e[%dB", $dy } : $dy < 0 ? do { printf "\e[%dA", -$dy } : do {};
}

sub save_cursor {
    my $self = shift;
    print "\e[s";
}

sub restore_cursor {
    my $self = shift;
    print "\e[u";
}

sub hide_cursor {
    my $self = shift;
    state $once = eval 'END { $self->show_cursor }';
    print "\e[?25l";
}
sub show_cursor {
    my $self = shift;
    print "\e[?25h";
}

sub clear_screen {
    my $self = shift;
    print "\e[1J";
    print "\e[1;1H";
}

sub get_frog {
    my $self = shift;
    my $n    = shift;
    return $self->_frogs->[$n];
}

sub set_frog {
    my ( $self, $index, $frog )= @_;
    $self->_frogs->[$index] = $frog;
}

sub frog_on_stone {
    my ( $self, $stone_n ) = @_;
    foreach my $n ( 0..$self->frog_number-1 ){
        my $frog = $self->get_frog($n);
        return $frog if $frog->stone_index == $stone_n;
    }
    my $null_frog = Games::FrogJump::Frog->new();
    return $null_frog;
}

sub jump_frog_left {
    my ( $self, $frog, $step ) = @_;
    $frog->stone_index($frog->stone_index - $step);

    my $stopx = $frog->x - $step * ( $self->stone_width + $self->stone_gap );

    my $animation_x = Games::FrogJump::Animation->new(
        name     => 'x',
        duration => 0.5,
        obj      => $frog,
        attr     => 'set_x',
        snapshot => [$frog->x, int(($frog->x + $stopx) / 2), $stopx],
        );
    my $animation_y = Games::FrogJump::Animation->new(
        name     => 'y',
        duration => 0.5,
        obj      => $frog,
        attr     => 'set_y',
        snapshot => [$frog->y, $frog->y - $step, $frog->y],
        );
    $self->add_animation($animation_x);
    $self->add_animation($animation_y);

}

sub jump_frog_right {
    my ( $self, $frog, $step ) = @_;
    $frog->stone_index($frog->stone_index + $step);

    my $stopx = $frog->x + $step * ( $self->stone_width + $self->stone_gap);

    my $animation_x = Games::FrogJump::Animation->new(
        name     => 'x',
        duration => 0.5,
        obj      => $frog,
        attr     => 'set_x',
        snapshot => [$frog->x, int(($frog->x + $stopx) / 2), $stopx],
        );
    my $animation_y = Games::FrogJump::Animation->new(
        name     => 'y',
        duration => 0.5,
        obj      => $frog,
        attr     => 'set_y',
        snapshot => [$frog->y, $frog->y - $step, $frog->y],
        );
    $self->add_animation($animation_x);
    $self->add_animation($animation_y);
}

sub _build__frogs {
    my $self = shift;
    [ 1..$self->frog_number ];
}


sub set_current_frog {
    my ( $self, $frog ) = @_;
    $self->current_frog->unactive if $self->current_frog ne '';
    $self->current_frog($frog);
    $frog->active;
}



1;


lib/Games/FrogJump/Input.pm
package Games::FrogJump::Input;
use 5.012;
use strict;
use warnings;

use if $^O eq "MSWin32", "Win32::Console::ANSI";
use Term::ReadKey;
use Time::HiRes;

eval 'END { ReadMode "normal" }';
eval 'ReadMode "cbreak"';


sub read_key {
	state @keys;

	if (@keys) {
		return shift @keys;
	}

	my $char;
	my $packet = '';
	while (defined($char = ReadKey -1)) {
		$packet .= $char;
	}

	while ($packet =~ m(
		\G(
			\e \[          # CSI
			[\x30-\x3f]*   # Parameter Bytes
			[\x20-\x2f]*   # Intermediate Bytes
			[\x40-\x7e]    # Final Byte
		|
			.              # Otherwise just any character
		)
	)gsx) {
		push @keys, $1;
	}
	return shift @keys;
}

sub key_to_cmd {
    my $key = shift;
    my $cmd = '';
    if ( $key eq "\e[C" ){
        $cmd = 'right';
    }
    if ( $key eq "\e[D" ){
        $cmd = 'left';
    }
    if ( $key eq " " ){
        $cmd = 'jump';
    }
    if ( $key eq 'q' ){
        $cmd = 'quit';
    }
    if ( $key eq 'r' ){
        $cmd = 'restart';
    }
    return $cmd;
}
1;


lib/Games/FrogJump/Animation.pm
package Games::FrogJump::Animation;
use 5.012;

use Moo;

has name          => is => 'rw', default => undef;
has current_frame => is => 'rw', default => 0;
has duration      => is => 'rw', default => 0;
has obj           => is => 'rw', default => undef;
has attr          => is => 'rw', default => undef;
has snapshot      => is => 'rw', default => undef;

sub frame_count {
    my $self = shift;
    return int($self->duration / $Games::FrogJump::FRAME_TIME);
}

sub snapshot_frame_count {
    my $self = shift;
    return int($self->frame_count / @{$self->snapshot});
}

sub end {
    my $self = shift;
    return $self->current_frame >= $self->frame_count ? 1 : 0;
}

sub update {
    my $self = shift;
    my $obj  = $self->obj;
    my $attr = $self->attr;
    my $snap_index = int($self->current_frame / $self->snapshot_frame_count);
    $obj->$attr($self->snapshot->[$snap_index]) if $snap_index < @{$self->snapshot};
    $self->current_frame($self->current_frame + 1);
}
1;


lib/Games/FrogJump/Game.pm
package Games::FrogJump::Game;
use 5.012;
use Moo;

use Storable;
use File::Spec::Functions;
use File::HomeDir;

extends 'Games::FrogJump::Board';

has init_directions   => is => 'ro' => default => sub { [ 'right', 'right', 'right', 'left', 'left', 'left' ] };
has target_directions => is => 'ro' => default => sub { [ 'left', 'left', 'left', 'null', 'right', 'right', 'right' ] };

sub init {
    my $self = shift;
    my $directions = $self->init_directions;
    foreach my $index ( 0..$self->frog_number - 1 ){
        my $frog = Games::FrogJump::Frog->new(
            direction    => $directions->[$index],
            stone_index  => $index < 3 ? $index : $index + 1,
            bg_color     => $index < 3 ? 'on_yellow' : 'on_green',
            ansi         => $index < 3 ? '_--'  : '--_',
            x            => $index < 3 ?
                            $self->border_width + $self->padding + ($self->stone_width - $self->frog_width) / 2 + $index * ( $self->stone_width + $self->stone_gap ) :
                            $self->border_width + $self->padding + ($self->stone_width - $self->frog_width) / 2 + ($index + 1) * ( $self->stone_width + $self->stone_gap ),
            y            => $self->content_height - 1,
            oldx         => $index < 3 ?
                            $self->border_width + $self->padding + ($self->stone_width - $self->frog_width) / 2 + $index * ( $self->stone_width + $self->stone_gap ) :
                            $self->border_width + $self->padding + ($self->stone_width - $self->frog_width) / 2 + ($index + 1) * ( $self->stone_width + $self->stone_gap ),
            oldy         => $self->content_height - 1,
            );
        $self->set_frog( $index, $frog );
    }
    $self->set_current_frog($self->get_frog(0));
    $SIG{INT} = sub { $self->draw_quit; exit 1 };
    $SIG{__DIE__} = sub { $self->draw_quit; exit 1};
    $self->clear_screen;
}

sub act {
    my $self = shift;
    my $cmd  = shift;
    if ( $cmd eq 'left' ){
        $self->move_left;
    }
    if ( $cmd eq 'right' ){
        $self->move_right;
    }
    if ( $cmd eq 'jump' ){
        $self->jump;
    }
}
sub jump {
    my ( $self ) = @_;
    my $current_frog = $self->current_frog;
    my $current_stone = $current_frog->stone_index;
    my $direction = $current_frog->direction;
    if ( $direction eq 'right' ){
        $self->alarm_no_jump and return if $current_stone == $self->stone_number - 1;
        my $next_frog = $self->frog_on_stone($current_stone + 1);
        if ( $next_frog->direction eq 'null' ){
            $self->jump_frog_right($current_frog, 1);
            return;
        }
        $self->alarm_no_jump and return if $current_stone == $self->stone_number - 2;
        my $next_next_frog = $self->frog_on_stone($current_stone + 2);
        if ( $next_next_frog->direction eq 'null' ){
            $self->jump_frog_right($current_frog, 2);
            return;
        }
        $self->alarm_no_jump and return;
    }
    if ( $direction eq 'left' ){
        $self->alarm_no_jump and return if $current_stone == 0;
        my $next_frog = $self->frog_on_stone($current_stone - 1);
        if ( $next_frog->direction eq 'null' ){
            $self->jump_frog_left($current_frog, 1);
            return;
        }
        $self->alarm_no_jump and return if $current_stone == 1;
        my $next_next_frog = $self->frog_on_stone($current_stone - 2);
        if ( $next_next_frog->direction eq 'null' ){
            $self->jump_frog_left($current_frog, 2);
            return;
        }
        $self->alarm_no_jump and return;
    }
}

sub move_left {
    my ( $self ) = @_;
    my $current_stone = $self->current_frog->stone_index;
    if( $current_stone == 0 ){
        return;
    }
    my $next_frog = $self->frog_on_stone($current_stone - 1);
    if ($next_frog->direction eq 'null' ) {
        if ( $current_stone == 1 ){
            return;
        }
        else{
            $next_frog = $self->frog_on_stone($current_stone - 2);
        }
    }
    $self->set_current_frog( $next_frog );
}

sub move_right {
    my ( $self ) = @_;
    my $current_stone = $self->current_frog->stone_index;
    if( $current_stone == $self->stone_number - 1 ){
        return;
    }
    my $next_frog = $self->frog_on_stone($current_stone + 1);
    if ($next_frog->direction eq 'null' ) {
        if ( $current_stone == $self->stone_number - 2 ){
            return;
        }
        else{
            $next_frog = $self->frog_on_stone($current_stone + 2);
        }
    }
    $self->set_current_frog( $next_frog );
}

sub alarm_no_jump {
    my $self = shift;
    my $animation = Games::FrogJump::Animation->new(
        name     => 'alarm_no_jump',
        duration => 0.5,
        obj      => $self,
        attr     => 'border_color',
        snapshot => ['on_white', $self->border_color],
        );
    $self->add_animation($animation);
    return 1;
}

sub win {
    my $self = shift;
    foreach my $stone ( 0..$self->stone_number-1 ){
        return 0 if $self->target_directions->[$stone] ne $self->frog_on_stone($stone)->direction;
    }
    return 1;
}

sub lose {
    my $self = shift;
    return 0;
}

sub restart {
    my $self = shift;
    $self->move_cursor(0, $self->border_height + $self->content_height);
    $self->clear_screen;
}
1;
